### ISSUE TITLE
Simple improvements to password validation

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#9150]
We previously tried to implement a "password strength indicator" into Ghost (ref #5309), but got stuck on the libraries being too big. However, in abandoning strength indication, we've totally missed out on adding very simple rules to prevent the most obvious and heinous of password mistakes.

We want to both increase the minimum length for passwords, and add a couple of very simple rules. 

The most important thing about these new validation rules, is that they **MUST** not interfere with existing users being able to login or save their profile. The new validation rules must also not prevent old blog content from being imported. They should ensure that a user creating an account, or changing a password, get validated more strongly.

The first 3 tasks go together, to cover making a single change to these rules and then ensuring that it doesn't impact anything it shouldn't. Once that is done, each of the following rules can be added one-by-one. 

**Increase minimum length:**

- [x] Increase password minimum from 8 to 10 characters (ensure it is messaged properly)
- [x] Testing to ensure this does **not** impact on saving user profiles, imports or any other user-based action where the password is not being changed.
- [x] Improve importer logic, see notes below

**Add simple rules:**

- [x] Design note: adding rules will require a new validation error message, this will need to be run past @JohnONolan 
- [x] Disallow obviously bad passwords: `1234567890`, `qwertyuiop`, `asdfghjkl;` and `abcdefghij`.
- [x] Disallow passwords that contain the words "password" or "ghost"
- [x] Disallow passwords that match the user's email address
- [x] Disallow passwords that match the blog domain
- [x] Add a very simple entropy measure with a cut off e.g. if 50% or more of characters are the same - this prevents 'aaaaaaaaaa' and 'ababababab' style passwords.

**Advanced changes (these are for a new issue, later):**
- Research an advanced entropy measure, with a score system
- Maybe add a larger library, but using lazy loading (this is now possible)
- Add a signal strength indicator to the UI

Ref material: https://blog.codinghorror.com/password-rules-are-bullshit/

Notes on Import rules:

At the moment, our import rules for passwords are a bit weird. There are 2 modes:

1. "importPersistUser" mode, is a special mode in which users should be persisted exactly as they were. No validations should be run.
2. "normal" mode, in which we expect to be importing a JSON file which is an original export, but it _may_ also be user generated. Here, validations should run but the outcome is slightly different...

Validating the field should be a case of: Check if the contents of the field is a hash:
- If it is a hash, don't need to validate, re-hash-the-hash, store that. 
- If it is _not_ a hash, all the normal password validation rules should be run, then hash.

@kirrg001 can you confirm this logic please?

**Comment by AileenCGN** [comment]
@erisDS question regarding this:

> If it is not a hash, all the normal password validation rules should be run, then hash.

Before, the length validation for passwords during the import process never fired, because of [this line](https://github.com/TryGhost/Ghost/blob/master/core/server/data/importer/importers/data/users.js#L46), which set **every** password to a 50 characters long UID.

To implement those changes and properly detect hashed passwords during the import process, I removed this line. 

What happens now is, that non-hashed passwords _may_ fail the validation for the required length and therefore the whole import process fails. Is this the required result?

Valid passwords get hashed and stored as desired, but only if no validation error in the import file occurs.

It looks like this:
![image](https://user-images.githubusercontent.com/8037602/31715472-368c6976-b42e-11e7-8cd2-71b2a191c428.png)

**Comment by AileenCGN** [comment]
One more question regarding the additional rules: Those rules shall take effect for signups, pw changes, pw resets (all of them mostly client side), **and** the importer?

**Comment by ErisDS** [comment]
Sorry - I've confused things by over-specifying & I missed the fact that we set the password to be a random password during normal imports (which is desired behaviour) by explicitly setting it to be a uid, I thought we did it by rehashing the hash...

Going back to problems first, not solutions, the problems with the importer logic I want to solve are:
1. That validations get run on hashes, even though they can never fail. They should not be run at all. Reason being that as we add more rules around passwords, the hash (or our force-set random password) could fail the validations, but that's incorrect.
2. In `importPersistUser` mode we assume, rather than check that we are copying a bcrypt hash across. It's a fairly safe assumption, but it would be safer to only do the copy IF the password is a bcrypt hash, and if it isn't fall back to our normal import behaviour of overwriting the password and locking the account.
3. I think from our discussions it's clear that the random-password-setting-code in the importer is hard to find. Plus we have 2 distinct places in the code base where we do different logic based on the `importPersistUser` flag. These things are a bit confusing.

^ Task 3 is solve these 3 problems however you feel works best 😁

**Comment by AileenCGN** [comment]
> Disallow passwords that match the blog domain

I think it also makes sense to disallow passwords that match the blog title?

**Comment by AileenCGN** [comment]
### Update:

> Add a very simple entropy measure with a cut off e.g. if 50% or more of characters are the same - this prevents 'aaaaaaaaaa' and 'ababababab' style passwords.

Regarding this last task that's in scope of this PR, I wasn't able to find anything that does only that. Any bigger entropy measure library does support it. As a first step towards safer passwords, I suggest, that we solve the last task within the next issue, which then includes the advanced changes as describes in the spec.

The PR for the client side, that includes the new simple rules is this one: TryGhost/Ghost-Admin#899
The PR for the server side, is this one: #9171

**Comment by AileenCGN** [comment]
To specify what I wrote in my comment above:

The task was to add a very simple entropy measure with a cut off when e. g. 50% or more characters are the same (e. g. 'aaaaaaaaaa' or 'ababababab').

**I couldn't find a solution for *only* that**, means I would either 

   1. code it myself (not done with 3 lines of code, fwik), or
   2. wait with this last task until we go ahead with the advanced password validation tasks and use 3rd party libraries.

Those advanced password validation tasks include the research for an advanced entropy measure library. I did a bit of research already for that (but not fully, as I was mainly concentrating on the simple password rules, which work without 3rd party libraries) and came over the following libraries:
- [zxcvbn](https://www.npmjs.com/package/zxcvbn): supports detecting repeats as documented, will give a `score` of `0` = super weak, therefore preventable
- [tai-password-strength](https://www.npmjs.com/package/tai-password-strength): does not support it explicitly, but will score these passwords as 'VERY_WEAK' which we could prevent to use then
- [passwd-strength](https://www.npmjs.com/package/passwd-strength): wouldn't recommend it, a password like `ababababab` would result in a score, that this package considers as not too weak

I also stumbled over more libraries, but they are not useful for our requirements, as they only test for given options like 'must have one uppercase', 'min-length', etc: [password-validator](https://www.npmjs.com/package/password-validator), [fnbpasswordvalidator](https://www.npmjs.com/package/fnbpasswordvalidator).

Again, I didn't dive too much into the overall research, as I was only looking for a solution for the task mentioned above.

**Comment by AileenCGN** [comment]
@ErisDS Sorry for the confusion. I updated the PRs with the last rule. Guess, I just needed a good sleep so the idea of how to solve this would occur in my dreams. 😔

**Comment by AileenCGN** [comment]
Interesting addition to the mentioned article on [codinghorror](https://blog.codinghorror.com/password-rules-are-bullshit/): https://paul.reviews/passwords-why-using-3-random-words-is-a-really-bad-idea/

**Comment by kirrg001** [comment]
@AileenCGN Can be closed?

**Comment by AileenCGN** [comment]
@kirrg001 Yes, but shall we add a "later" label, so we won't forget about the "Advanced changes (these are for a new issue, later)"?

**Comment by ermyril** [comment]
Hello!

Sorry for bumping an old issue, it's more like a feature request, but please add more descriptive validation errors. It's definitely not obvious what's wrong when an application just tells you "Sorry, you cannot use an insecure password"
when you're just typed something like "GhostinglyCrypticGibberish8*5%"

**Comment by huttj** [comment]
Just ran into the same issue and am disinclined from signing up for the service when I'm told that my long password is somehow "insecure."

**Comment by JohnONolan** [comment]
Sorry for the ambiguous error message! Usually shows up if you're using "ghost" as part of the password - which I agree is confusing. We'll update
